File: sources/opencv-samples/12_ImageSegmentation/main.py
import cv2 as cv
import numpy as np
import sys
import signal
import argparse
from time import time


global has_to_break


def sigint_handler(signal, frame):
    global has_to_break
    has_to_break = True


def getTime():
	return time()


def getTimeDelta(t):
	t = 1000 * (getTime() - t)
	return round(1000 * t) / 1000


def main():
	global has_to_break
	has_to_break = False

	args = parser.parse_args()
	
	path = cv.samples.findFile(args.img_path)
	
	if not path:
		sys.exit(f"Image {args.img_path} not found")
	
	image = cv.imread(path)

	if image is None:
		sys.exit(f"Could not read the image {path}")

	print("")
	print("press Left Arrow to decrease the threshold value")
	print("press Rigth Arrow to increase the threshold value\n")
	print("press ESC to quit\n")

	k = 0

	window1 = "Image"
	window2 = "Result"

	# Pour la binarisation de l'image
	threshold_value = 100   # seuil de binarisation

	t = 0.0
	has_to_compute = True

	# Convertir en niveaux de gris
	gray = cv.cvtColor(image, cv.COLOR_BGR2GRAY)

	cv.imshow(window1, image)

	while True:

		if has_to_compute:
			has_to_compute = False

			print("")
			print(f"threshold : {threshold_value}")

			t = getTime()

			_, binary = cv.threshold(gray, threshold_value, 255, cv.THRESH_BINARY)

			# Inverser l'image binaire pour obtenir les régions de l'ombre
			binary_inverse = cv.bitwise_not(binary)

			# Remplacer les pixels de l'ombre par des pixels de l'image d'origine
			image_no_shadow = cv.bitwise_or(image, image, mask=binary_inverse)

			print(f"duration: {getTimeDelta(t)} ms")

			cv.imshow(window2, image_no_shadow)

		k = cv.waitKey(0) & 0xFF

		if has_to_break or k == 27:
			break

		elif k == 81:  # Left Arrow
			if threshold_value > 0:
				threshold_value -= 1
				has_to_compute = True

		elif k == 83:  # Right Arrow
			if threshold_value < 255:
				threshold_value += 1
				has_to_compute = True

	cv.waitKey(0)
	cv.destroyAllWindows()

if __name__ == '__main__':
    signal.signal(signal.SIGINT, sigint_handler)
    parser = argparse.ArgumentParser(description="Image Segmentation main.py CLI")
    parser.add_argument("img_path", type=str, help="image path to open")
    main()

File: sources/opencv-samples/12_ImageSegmentation/main.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <csignal>

using namespace std;
using namespace cv;

bool hasToBreak = false;

void signal_handler(int signal)
{
    hasToBreak = true;
}

double getTime()
{
    return (double)getTickCount();
}

double getTimeDiff(double time_ori)
{
    return 1000 * (getTime() - time_ori) / getTickFrequency();
}

int main(int argc, char** argv) 
{
    // Install a signal handler
    std::signal(SIGINT, signal_handler);

    if (argc != 2)
    {
        printf("Usage: ./main.out <image_path>");
        return -1;
    }

    Mat image = imread(argv[1]);

    if (!image.data)
    {
        cout << "Could not read the image " << argv[1] << endl;
        return -1;
    }

    printf("\n");
    printf("press Left Arrow to decrease the threshold value\n");
    printf("press Rigth Arrow to increase the threshold value\n\n");
    printf("press ESC to quit\n\n");

    char k;

    string window1 = "Image";
    string window2 = "Result";

    Mat gray;

    // Pour la binarisation de l'image
    int threshold_value = 100;   // seuil de binarisation

    double t;
    bool hasToCompute = true;

    cvtColor(image, gray, cv::COLOR_BGR2GRAY);

    imshow(window1, image);

    while (1)
    {
        if (hasToCompute)
        {
            hasToCompute = false;
            cout << endl;
            cout << "threshold : " << threshold_value << endl;
            
            t = getTime();
            
            // Appliquer un seuillage binaire pour obtenir un masque de l'ombre
            Mat binary;
            threshold(gray, binary, threshold_value, 255, cv::THRESH_BINARY);

            // Inverser l'image binaire pour obtenir les régions de l'ombre
            Mat binary_inverse;
            bitwise_not(binary, binary_inverse);

            // Remplacer les pixels de l'ombre par des pixels de l'image d'origine
            Mat image_no_shadow;
            bitwise_or(image, image, image_no_shadow, binary_inverse);

            cout << "duration : " << getTimeDiff(t) << " ms" << endl;

            imshow(window2, image_no_shadow);
        }

        k = (char)waitKey(0);

        if (hasToBreak ||                                                   // due to Ctrl+C
            k == 27 ||                                                      // ESC key
            cv::getWindowProperty(window1, WND_PROP_AUTOSIZE) == -1 ||
            cv::getWindowProperty(window2, WND_PROP_AUTOSIZE) == -1)        // window is closed
        {
            break;
        }
        else if (k == 81)  // Left Arrow
        {
            if (threshold_value > 0)
            {
                threshold_value--;
                hasToCompute = true;
            }
        }
        else if (k == 83)  // Right Arrow
        {
            if (threshold_value < 255)
            {
                threshold_value++;
                hasToCompute = true;
            }
        }
    }
    destroyAllWindows();
    return 0;
}
